### PR TITLE
Api gateway

### PR DESCRIPTION
**Pull request by Jorrit05**

### CODE REVIEW

File: charts/api-gateway/templates/sample.yaml
Comment (Jorrit05) [inline charts/api-gateway/templates/sample.yaml:1]: remove if not required (clean code principal)

File: charts/api-gateway/values.yaml
Comment (Jorrit05) [inline charts/api-gateway/values.yaml:1]: hostPath can probably be removed

File: charts/namespaces/templates/namespaces.yaml [hunk 1]
```diff
@@ -119,4 +119,26 @@ metadata:
...
+metadata:
+  name:  api-gateway
+  annotations:
+    "helm.sh/resource-policy": keep
+    "linkerd.io/inject" : enabled
+    "config.linkerd.io/trace-collector": collector.linkerd-jaeger:55678 # or 14268?
+
+---
+
+apiVersion: v1
```
Comment (Jorrit05) [inline charts/namespaces/templates/namespaces.yaml:137]: this secret is interesting. If you look at 'values.yaml' it will set a default password. I did this for the CloudLab installation and will overwrite the secrets later. 

Hoewver, running locally it should give connections issues to rabbitMQ. I am honestly not sure why it is running on my machine witout errors...

Keep this secret here though, this just my thought process running

File: go/cmd/api-gateway/config_prod.go [hunk 2]
```diff
@@ -0,0 +1,14 @@
+//go:build !local
+// +build !local
+
+package main
+
+import "go.uber.org/zap"
+
+var serviceName = "orchestrator"
```
Comment (Jorrit05) [inline go/cmd/api-gateway/config_prod.go:8]: this needs to change for RabbitMQ to work

File: go/cmd/api-gateway/config_local.go [hunk 3]
```diff
@@ -0,0 +1,10 @@
+//go:build local
+// +build local
+
+package main
+
+import "go.uber.org/zap"
+
```
Comment (Jorrit05) [inline go/cmd/api-gateway/config_local.go:7]: add servicename here as well

File: go/cmd/api-gateway/consume.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+	ctx, span, err := lib.StartRemoteParentSpan(ctx, serviceName+"/func: handleIncomingMessages", grpcMsg.Traces)
+	if err != nil {
+		logger.Sugar().Warnf("Error starting span: %v", err)
+	}
+	defer span.End()
+
+	logger.Sugar().Debugw("Type:", "MessageType", grpcMsg.Type)
+
+	switch grpcMsg.Type {
+	case "requestApprovalResponse":
```
Comment (Jorrit05) [inline go/cmd/api-gateway/consume.go:22]: the case should mostly match the struct type, which it does not do here:
		var requestApproval pb.RequestApproval

Maybe good to draw a simplified view of the message flow? Than we can discuss message flow more easily

So how would we do it now if the request was not accepted? 

Maybe it is good to change the message type  "RequestApproval" to the one you mentio here:
"requestApprovalResponse" which then might contain an error field

File: go/cmd/api-gateway/consume.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Jorrit05/DYNAMOS/pkg/lib"
+	pb "github.com/Jorrit05/DYNAMOS/pkg/proto"
+)
+
+func handleIncomingMessages(ctx context.Context, grpcMsg *pb.SideCarMessage) error {
```
Comment (Jorrit05) [inline go/cmd/api-gateway/consume.go:11]: General comment: we should probably add short 'functional' comment lines to explain a bit more of why a function is there and doing something/

(Although I should've done that more in my thesis :) )

File: go/cmd/api-gateway/config_prod.go [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+var port = ":8080"
```
Comment (aleandromifsud) [inline go/cmd/api-gateway/config_prod.go:9]: Should the port not be 8080 because of the orchestrator?

File: configuration/grafana/other.json
Comment (aleandromifsud) [inline configuration/grafana/other.json:1]: Ill revert the changes in this file, they were accidental

File: go/Makefile [hunk 7]
```diff
@@ -7,7 +7,7 @@ prepare:
 proto:
 	protoc -I ../proto-files --go_out=./pkg/proto --go_opt=paths=source_relative --go-grpc_out=./pkg/proto --go-grpc_opt=paths=source_relative ../proto-files/*.proto
 
-$(targets): prepare proto
```
Comment (aleandromifsud) [inline go/Makefile:10]: Ill also revert this

File: go/cmd/orchestrator/main.go [hunk 8]
```diff
@@ -83,6 +85,7 @@ func main() {
 	apiMux.Handle("/policyEnforcer", &ochttp.Handler{Handler: agreementsHandler(etcdClient, "/policyEnforcer")})
 	apiMux.Handle("/policyEnforcer/", &ochttp.Handler{Handler: agreementsHandler(etcdClient, "/policyEnforcer")})
 
+	// TODO Replace with rabbitMQ receive
 	apiMux.Handle("/requestapproval", &ochttp.Handler{Handler: requestApprovalHandler()})
```
Comment (aleandromifsud) [inline go/cmd/orchestrator/main.go:89]: This can probably be removed now

File: go/cmd/api-gateway/consume.go (hunk 4 above)
Comment (aleandromifsud) [reply go/cmd/api-gateway/consume.go:22]: That makes sense, now we have the two sides of communication tho, 

so 
(1) Api Gateway -- requestApprovalRequest (RequestApproval) --> Orchestrator 
and then
(2) Orchestrator --- requestApprovalResponse (AcceptedDataRequest) --> Api Gateway

i think changing the RequestApproval might be a bit confusing since it is the initial request, I do however agree that it does not make sense to have AccepetedDataRequest as the response since it might contain errors.

So maybe we should change AcceptedDataRequest to RequestApprovalResponse? and that can include the errors?

what do you think?

File: go/cmd/api-gateway/config_prod.go (hunk 6 above)
Comment (aleandromifsud) [reply go/cmd/api-gateway/config_prod.go:9]: Actually now that i think about it, they are on different pods so the same port should work fine

File: go/cmd/api-gateway/consume.go (hunk 4 above)
Comment (Jorrit05) [reply go/cmd/api-gateway/consume.go:25]: ah right, I see now that in my comment I messed it up. 

So yeah datatype AcceptedDataRequest should change to -> requestApprovalResponse

And this one requestApprovalRequest, might (let's discuss)  be shortened to simply 'requestApproval' its simply passing it on
